---
<div style="text-align: center;">
<h1>The University of North Carolina at Chapel Hill</h1>
<h1>Comp 555 BioAlgorithms - Spring 2022</h1>
<h1 style="font-size: 250%;">Problem Set #6</h1>
<h1>Issued Thursday, 4/11/2022; Due Thursday, 4/21/2022</h1>
</div>

**Homework Information:** Some of the problems are probably too long to be done the night before the due date, so plan accordingly. Feel free to get help from others, but **the work you submit in should be your own.**

**Warning:** This notebook has been annotated with metadata so that it can be uploaded to the grading system. It is very important that you enter your answers in the provided cells. You can add extra cells to explore approaches, but only the provided cell can and will be graded. Thus, if you delete a cell and add a replacement, there is a possiblity that your problem will not be graded. If you ever need to start over, you should download a new version of the problem set and transfer your solutions to it.

In [1]:
# Replace the following string values with the requested information
class Student:
    first = "Owen"
    last = "McCadden"
    onyen = "owenmc"
    pid = "730314080"

---
**Problem #1:** Assuming an *"Unfair-Casino"* with the same rules as discussed in class (Rolls can be either from a fair coin with an equal proabliity of landing Heads or Tails, or a biased coin that lands Heads with probability 0.75 and Tails with probability 0.25, meanwhile the dealer swaps coins with a probability of 0.1). Suppose that the following series of coin flips are observed:

        HTHHHHHHHHHTHTTTHTHTHHHHTTHHTHTTTHHTHHHHHHHTHHHTHHTTTHHHHTTHHHHHHHTTHTHHHHHHHTTH

What series of coin states maximizies the likelihood of this pattern of rolls and what is it's probability? What is the most likely series of coin-states that generates this pattern of rolls where the dealer ends up holding the other coin, and what is it's probability?

Lastly, after reviewing the casino's cameras, it is deterimed that the actual series of coin states was:

        FFFBBBBBBBBBBBBBBBFFBBBBBBBBBBBBBBBBBBBBBBBBBBBBFFFFFFFFFFFFBBBFBBBBBBBBBBBBBBBB
        
What was the probaility that this series of coin states occured and produced the given pattern of rolls?

In [100]:
sigma = ['H', 'T']
Q = ['F', 'B']
A = np.array([[0.9, 0.1],[0.1, 0.9]])
E = np.array([[0.5, 0.5],[0.25, 0.75]])


observed = "HTHHHHHHHHHTHTTTHTHTHHHHTTHHTHTTTHHTHHHHHHHTHHHTHHTTTHHHHTTHHHHHHHTTHTHHHHHHHTTH"

In [119]:
import numpy as np

def viterbi(A, C, B, O):
    """Viterbi algorithm for solving the uncovering problem

    Notebook: C5/C5S3_Viterbi.ipynb

    Args:
        A (np.ndarray): State transition probability matrix of dimension I x I
        C (np.ndarray): Initial state distribution  of dimension I
        B (np.ndarray): Output probability matrix of dimension I x K
        O (np.ndarray): Observation sequence of length N

    Returns:
        S_opt (np.ndarray): Optimal state sequence of length N
        D (np.ndarray): Accumulated probability matrix
        E (np.ndarray): Backtracking matrix
    """
    I = A.shape[0]    # Number of states
    N = len(O)  # Length of observation sequence

    # Initialize D and E matrices
    D = np.zeros((I, N))
    E = np.zeros((I, N-1)).astype(np.int32)
    D[:, 0] = np.multiply(C, B[:, O[0]])

    # Compute D and E in a nested loop
    for n in range(1, N):
        for i in range(I):
            temp_product = np.multiply(A[:, i], D[:, n-1])
            D[i, n] = np.max(temp_product) * B[i, O[n]]
            E[i, n-1] = np.argmax(temp_product)

    # Backtracking
    S_opt = np.zeros(N).astype(np.int32)
    S_opt[-1] = np.argmax(D[:, -1])
    for n in range(N-2, -1, -1):
        S_opt[n] = E[int(S_opt[n+1]), n]

    return S_opt, D, E

# Define model parameters
A = np.array([[0.9, 0.1], 
              [0.1, 0.9]])

C = np.array([0.5, 0.5])

B = np.array([[0.5, 0.5], 
              [0.25, 0.75]])

observed = "HTHHHHHHHHHTHTTTHTHTHHHHTTHHTHTTTHHTHHHHHHHTHHHTHHTTTHHHHTTHHHHHHHTTHTHHHHHHHTTH"
O = np.array(list([{'H': 1, 'T':0}[char] for char in observed]))
#O = np.array([1]).astype(np.int32)
#O = np.array([1, 2, 0, 2, 2, 1]).astype(np.int32)

# Apply Viterbi algorithm
S_opt, D, E = viterbi(A, C, B, O)
S_opt = [{1:"H", 0:"T"}[char] for char in S_opt]
#
print('Observation sequence:   O = ', O)
print('Optimal state sequence: S = ', S_opt)
np.set_printoptions(formatter={'float': "{: 7.4f}".format})
print('D =', D, sep='\n')
np.set_printoptions(formatter={'float': "{: 7.0f}".format})
print('E =', E, sep='\n')

Observation sequence:   O =  [1 0 1 1 1 1 1 1 1 1 1 0 1 0 0 0 1 0 1 0 1 1 1 1 0 0 1 1 0 1 0 0 0 1 1 0 1
 1 1 1 1 1 1 0 1 1 1 0 1 1 0 0 0 1 1 1 1 0 0 1 1 1 1 1 1 1 0 0 1 0 1 1 1 1
 1 1 1 0 0 1]
Optimal state sequence: S =  ['H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H', 'H']
D =
[[ 0.2500  0.1125  0.0506  0.0228  0.0103  0.0046  0.0021  0.0009  0.0004
   0.0003  0.0002  0.0001  0.0001  0.0000  0.0000  0.0000  0.0000  0.0000
   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
   

---
**Problem #2:** For this problem you will need the genotypes in the file <a href="http://csbio.unc.edu/mcmillan/Comp555S22/PS06Data.csv" download="PS06Data.csv">"PS06Data.csv"</a>. This file has genotypes with chromosome and position for eight founder mouse strains (A/J, C57BL/6J, 129S1/SvImJ, NOD/ShiLtJ, NZO/HlLtJ, CAST/EiJ, PWK/PhJ, WSB/EiJ) followed by columns of new strains, one named for each student in Comp555. Look for your Comp555 username to identify your mouse strain and  infer a mosaic of founders that maximizes your strain's genotypes using the following assumptions. The likelihood of a genotyping error is 1%, and the probabily of a recombination between two genotypes can be modeled by an exponential function with a 50% probability of a change of founders when the genotype loci are separated by 20 Mbases. In space below print a line with each marker's position, your sample's genotype at that position, and the most likely founder for that genotype in your strain.    

In [20]:
fp = open("PS06Data.csv", 'r')
data = fp.read().split('\n')        # break file into lines
fp.close()
header = data.pop(0).split(',')     # First line is header
column = header.index("owenmc")

for i, line in enumerate(data):     # make a list from each row
    field = line.split(',')
    field[1] = int(field[1])        # convert position to integer
    data[i] = field[0:10] + [field[column]]

print(header[0:10] + [header[column]])
for i in range(0, 10):
    print("data[%d] = %s" % (i, data[i]))

['Chromosome', 'Position', 'A/J', 'C57BL/6J', '129S1/SvImJ', 'NOD/ShiLtJ', 'NZO/HlLtJ', 'CAST/EiJ', 'PWK/PhJ', 'WSB/EiJ', 'owenmc']
data[0] = ['13', 3561535, 'A', 'C', 'A', 'C', 'C', 'C', 'C', 'C', 'C']
data[1] = ['13', 4029227, 'A', 'A', 'A', 'A', 'G', 'G', 'A', 'A', 'A']
data[2] = ['13', 4752533, 'C', 'T', 'C', 'T', 'C', 'T', 'C', 'T', 'C']
data[3] = ['13', 5392085, 'T', 'G', 'T', 'T', 'G', 'T', 'G', 'T', 'G']
data[4] = ['13', 5945123, 'G', 'G', 'G', 'A', 'G', 'A', 'G', 'G', 'G']
data[5] = ['13', 6304055, 'G', 'A', 'G', 'G', 'G', 'G', 'G', 'G', 'G']
data[6] = ['13', 7088753, 'T', 'T', 'T', 'C', 'C', 'T', 'C', 'T', 'C']
data[7] = ['13', 7553047, 'A', 'A', 'A', 'C', 'C', 'A', 'C', 'C', 'C']
data[8] = ['13', 7999300, 'G', 'G', 'G', 'T', 'T', 'G', 'T', 'G', 'T']
data[9] = ['13', 8618531, 'T', 'T', 'T', 'C', 'C', 'C', 'C', 'C', 'C']


In [31]:
from math import exp, log10

Nstates = 8
prevpos = 1
state = [[(float(len(data)),i) for i in range(Nstates)]]   # (log(p), PathToHere)
for i in range(len(data)):
    # Count expected genotypes
    count = dict([(call, data[i][2:2+Nstates].count(call)) for call in "ACGTHN"])
    # Get the target genotype at this probe
    observed = data[i][-1]
    # Compute emission probability, assuming 1% error rate
    if (count[observed] == 0):
        emission = [1.0/Nstates for j in xrange(2,2+Nstates)]   # unexpected 
    else:
        emission = [0.99/count[data[i][j]] if data[i][j] == observed else 0.01/(Nstates - count[data[i][j]])
                    for j in range(2,2+Nstates)]
    # compute transition probability
    position = data[i][1]
    delta = position - prevpos
    prevpos = position
    stay = ((Nstates - 1.0)*exp(-delta/23604450.50) + 1.0)/Nstates
    switch = (1.0 - stay)/(Nstates - 1.0)
    # update state probailities for all paths leading to the ith state
    path = []
    for j in range(Nstates):
        choices = [(log10(emission[j])+(log10(stay) if (k==j) else log10(switch))+state[-1][k][0],k)
                   for k in range(Nstates)]
        path.append(max(choices))   # choices is a list of tuples of (score[i], from_whence_I_arrived[i])
    state.append(path)
print("Length of paths:", len(state))
print(state[-1])
print(state[-1][0][0])

Length of paths: 261
[(84.16240766244081, 0), (84.16240766244081, 1), (80.1711372732457, 2), (84.16240766244081, 3), (84.16240766244081, 4), (84.16240766244081, 5), (80.1711372732457, 6), (86.61094180229404, 7)]
84.16240766244081


In [32]:
# backtrack
path = state[-1]
maxi = 0
maxp = path[0][0]
for i in range(1,Nstates):
    if (path[i][0] > maxp):
        maxp = path[i][0]
        maxi = i
print(maxi, path[maxi], header[2+maxi])

for j in range(len(state)-2,-1,-1):
    data[j].append(header[2+maxi])
    maxi = state[j+1][maxi][1]

# for row in data:
#     print(row[1], row[-2], row[-1])

7 (86.61094180229404, 7) WSB/EiJ


---
**Problem #3:** Create an *anti-C57BL/6J* mouse as follows. For each genotype of C57BL/6J choose the alternate allele for the anti mouse. For example for the following genotypes:

    Chromosome,Position,A/J,C57BL/6J,129S1/SvImJ,NOD/ShiLtJ,NZO/HlLtJ,CAST/EiJ,PWK/PhJ,WSB/EiJ
    13,3561535,A,C,A,C,C,C,C,C
    13,4029227,A,A,A,A,G,G,A,A
    
The *C* allele of C57BL/6J at 13:3561535 would be imply an *A* in the *anti-C57BL/6J* mouse, and the *A* of 13:4029227 would imply a *G*. Once this anti mouse is constructed infer a mosaic of founders that maximizes the likelihood of the strain's genotypes using the same assumptions, and outputting in the same format as **Problem #1**.

In [76]:
fp = open("PS06Data.csv", 'r')
data = fp.read().split('\n')        # break file into lines
fp.close()
header = data.pop(0).split(',')     # First line is header
column = header.index("owenmc")
# while (len(data[-1].strip()) < 1):  # remove extra lines
#     data.pop()
for i, line in enumerate(data):     # make a list from each row
    field = line.split(',')
    field[1] = int(field[1])        # convert position to integer
    for base in field[2:10]:
        if base != field[3]:
            break
    data[i] = field[0:10] + [base]
    

for i in range(0, 10):
    print("data[%d] = %s" % (i, data[i]))

data[0] = ['13', 3561535, 'A', 'C', 'A', 'C', 'C', 'C', 'C', 'C', 'A']
data[1] = ['13', 4029227, 'A', 'A', 'A', 'A', 'G', 'G', 'A', 'A', 'G']
data[2] = ['13', 4752533, 'C', 'T', 'C', 'T', 'C', 'T', 'C', 'T', 'C']
data[3] = ['13', 5392085, 'T', 'G', 'T', 'T', 'G', 'T', 'G', 'T', 'T']
data[4] = ['13', 5945123, 'G', 'G', 'G', 'A', 'G', 'A', 'G', 'G', 'A']
data[5] = ['13', 6304055, 'G', 'A', 'G', 'G', 'G', 'G', 'G', 'G', 'G']
data[6] = ['13', 7088753, 'T', 'T', 'T', 'C', 'C', 'T', 'C', 'T', 'C']
data[7] = ['13', 7553047, 'A', 'A', 'A', 'C', 'C', 'A', 'C', 'C', 'C']
data[8] = ['13', 7999300, 'G', 'G', 'G', 'T', 'T', 'G', 'T', 'G', 'T']
data[9] = ['13', 8618531, 'T', 'T', 'T', 'C', 'C', 'C', 'C', 'C', 'C']


In [25]:
from math import exp, log10

Nstates = 8
prevpos = 1
state = [[(float(len(data)),i) for i in range(Nstates)]]   # (log(p), PathToHere)
for i in range(len(data)):
    # Count expected genotypes
    count = dict([(call, data[i][2:2+Nstates].count(call)) for call in "ACGTHN"])
    # Get the target genotype at this probe
    observed = data[i][-1]
    # Compute emission probability, assuming 5% error rate
    if (count[observed] == 0):
        emission = [1.0/Nstates for j in xrange(2,2+Nstates)]   # unexpected 
    else:
        emission = [0.99/count[data[i][j]] if data[i][j] == observed else 0.01/(Nstates - count[data[i][j]])
                    for j in range(2,2+Nstates)]
    # compute transition probability
    position = data[i][1]
    delta = position - prevpos
    prevpos = position
    stay = ((Nstates - 1.0)*exp(-delta/23604450.50) + 1.0)/Nstates
    switch = (1.0 - stay)/(Nstates - 1.0)
    # update state probailities for all paths leading to the ith state
    path = []
    for j in range(Nstates):
        choices = [(log10(emission[j])+(log10(stay) if (k==j) else log10(switch))+state[-1][k][0],k)
                   for k in range(Nstates)]
        path.append(max(choices))   # choices is a list of tuples of (score[i], from_whence_I_arrived[i])
    state.append(path)
print("Length of paths:", len(state))

Length of paths: 261


In [78]:
# backtrack
path = state[-1]
maxi = 0
maxp = path[0][0]
for i in range(1,Nstates):
    if (path[i][0] > maxp):
        maxp = path[i][0]
        maxi = i
print(maxi, path[maxi], header[2+maxi])

for j in range(len(state)-2,-1,-1):
    data[j].append(header[2+maxi])
    maxi = state[j+1][maxi][1]

for row in data:
    print(row[1], row[-2], row[-1])

2 (-13.933684392229354, 2) 129S1/SvImJ
3561535 A NZO/HlLtJ
4029227 G NZO/HlLtJ
4752533 C NZO/HlLtJ
5392085 T NOD/ShiLtJ
5945123 A NOD/ShiLtJ
6304055 G NOD/ShiLtJ
7088753 C NOD/ShiLtJ
7553047 C NOD/ShiLtJ
7999300 T NOD/ShiLtJ
8618531 C NOD/ShiLtJ
9636120 T NOD/ShiLtJ
10594297 T NOD/ShiLtJ
11491322 T NOD/ShiLtJ
12535416 C CAST/EiJ
13268590 A CAST/EiJ
14195930 C CAST/EiJ
14616277 A CAST/EiJ
15022703 C CAST/EiJ
15617153 G CAST/EiJ
16847837 G A/J
16935026 A A/J
17236701 T A/J
17868992 A A/J
18246094 T A/J
18626399 G A/J
18764095 G A/J
19162715 T A/J
19289945 C PWK/PhJ
19658269 A PWK/PhJ
20159611 T PWK/PhJ
20426383 A PWK/PhJ
22090835 C PWK/PhJ
23270280 C PWK/PhJ
23735565 T PWK/PhJ
25131524 T CAST/EiJ
25835082 A CAST/EiJ
26576136 A CAST/EiJ
27045522 T CAST/EiJ
28481382 G CAST/EiJ
28960442 G CAST/EiJ
29312977 C CAST/EiJ
29960408 G CAST/EiJ
30264993 C CAST/EiJ
30790623 C CAST/EiJ
31125127 C A/J
31277921 A A/J
31406941 G A/J
31406993 G A/J
32124864 G A/J
32641487 A A/J
32943624 C A/J
33071258 A 

---
**Problem #4:** How many reversal breakpoints does the permutation [8, 7, 10, 9, 6, 5, 4, 3, 2, 1] have? What does this imply about the lower bound on the number of reversals necessary to sort the given permutation?

In [41]:
def count_breakpoints(p):
    p.append(max(p)+1)
    p.insert(min(p)-1, 0)
    return sum([1 for i in range(len(p)-1) if abs(p[i]-p[i+1]) > 1])

In [42]:
seq = [8, 7, 10, 9, 6, 5, 4, 3, 2, 1]
count_breakpoints(seq)

4

---
**Problem #5:** How many reversals does the **GreedyReversalSort()** algorithm require to sort the given permutation? How many reversals does the **improvedBreakpointReversalSort()** algorithm require? Is this what you'd expected? Explain why?

In [80]:
def GreedyReversalSort(pi):
    t = 0
    for i in range(len(pi)-1):
        j = pi.index(min(pi[i:]))
        if (j != i):
            pi = pi[:i] + [v for v in reversed(pi[i:j+1])] + pi[j+1:]
            print("rho(%2d,%2d) = %s" % (i+1,j+1,pi))
            t += 1
    return t

print(GreedyReversalSort(seq))

rho( 1,10) = [1, 2, 3, 4, 5, 6, 9, 10, 7, 8]
rho( 7, 9) = [1, 2, 3, 4, 5, 6, 7, 10, 9, 8]
rho( 8,10) = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
3


In [12]:
def hasBreakpoints(seq):
    """ returns True if sequences is not strictly increasing by 1 """
    for i in range(1, len(seq)):
        if (seq[i] != seq[i-1] + 1):
            return True
    return False

def getStrips(seq):
    """ find contained intervals where sequence is ordered, and return intervals
    in as lists, increasing and decreasing. Single elements are considered
    decreasing. "Contained" excludes the first and last interval. """
    deltas = [seq[i+1] - seq[i] for i in range(len(seq)-1)]
    increasing = list()
    decreasing = list()
    start = 0
    for i, diff in enumerate(deltas):
        if (abs(diff) == 1) and (diff == deltas[start]):
            continue
        if (start > 0):
            if deltas[start] == 1:
                increasing.append((start, i+1))
            else:
                decreasing.append((start, i+1))
        start = i+1
    return increasing, decreasing

def pickReversal(seq, strips):
    """ test each decreasing interval to see if it leads to a reversal that
    removes two breakpoints, otherwise, return a reversal that removes only one """
    for i, j in strips:
        k = seq.index(seq[j-1]-1)
        if (seq[k+1] + 1 == seq[j]):
            # removes 2 breakpoints
            return 2, (min(k+1, j), max(k+1, j))
    # In the worst case we remove only one, but avoid the length "1" strips
    for i, j in strips:
        k = seq.index(seq[j-1]-1)
        if (j - i > 1):
            break
    return 1, (min(k+1, j), max(k+1, j))

def doReversal(seq,reversal):
    i, j = reversal
    return seq[:i] + [element for element in reversed(seq[i:j])] + seq[j:]

In [13]:
def improvedBreakpointReversalSort(seq, verbose=True):
    seq = [0] + seq + [max(seq)+1]                             # Extend sequence
    if (verbose):
        print("0: %s" % seq)
    N = 0
    while hasBreakpoints(seq):
        increasing, decreasing = getStrips(seq)
        if len(decreasing) > 0:                                # pick a reversal that removes a decreasing strip
            removed, reversal = pickReversal(seq, decreasing)
        else:
            removed, reversal = 0, increasing[0]               # No breakpoints can be removed
        if verbose:
            print("Strips:", increasing, decreasing)
            print("%d: %s  rho%s" % (removed, seq, reversal))
        seq = doReversal(seq,reversal)
        N += 1
    if verbose:
        print(seq, "Sorted")
    return N

In [14]:
seq = [8, 7, 10, 9, 6, 5, 4, 3, 2, 1]
GreedyReversalSort(seq)

rho( 1,10) = [1, 2, 3, 4, 5, 6, 9, 10, 7, 8]
rho( 7, 9) = [1, 2, 3, 4, 5, 6, 7, 10, 9, 8]
rho( 8,10) = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


3

In [15]:
improvedBreakpointReversalSort(seq)

0: [0, 8, 7, 10, 9, 6, 5, 4, 3, 2, 1, 11]
Strips: [] [(1, 3), (3, 5), (5, 11)]
1: [0, 8, 7, 10, 9, 6, 5, 4, 3, 2, 1, 11]  rho(3, 6)
Strips: [(4, 6)] [(1, 4), (6, 11)]
1: [0, 8, 7, 6, 9, 10, 5, 4, 3, 2, 1, 11]  rho(4, 7)
Strips: [] [(1, 5), (5, 7), (7, 11)]
1: [0, 8, 7, 6, 5, 10, 9, 4, 3, 2, 1, 11]  rho(5, 8)
Strips: [(6, 8)] [(1, 6), (8, 11)]
1: [0, 8, 7, 6, 5, 4, 9, 10, 3, 2, 1, 11]  rho(6, 9)
Strips: [] [(1, 7), (7, 9), (9, 11)]
1: [0, 8, 7, 6, 5, 4, 3, 10, 9, 2, 1, 11]  rho(7, 10)
Strips: [(8, 10)] [(1, 8), (10, 11)]
1: [0, 8, 7, 6, 5, 4, 3, 2, 9, 10, 1, 11]  rho(8, 11)
Strips: [] [(1, 9), (9, 11)]
1: [0, 8, 7, 6, 5, 4, 3, 2, 1, 10, 9, 11]  rho(1, 9)
Strips: [] [(9, 11)]
2: [0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 9, 11]  rho(9, 11)
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11] Sorted


8

Click [here to submit](http://csbio.unc.edu/mcmillan/index.py?run=PS.upload) your completed problem set